### PR TITLE
Add an optional verifier to crowdloan

### PR DESCRIPTION
**Pull request by xlc**
This will allow crowd loan owner to enforce any custom policy offchain for contributors by requiring a valid signature before accept a contribution.

TODOs:
- [x] docs
- [ ] test
- [x] think about replay attack

Based on #2166

### CODE REVIEW

**Review by gavofyork — APPROVED** [review]

File: runtime/common/src/crowdloan.rs [hunk 1]
```diff
@@ -334,9 +334,11 @@ decl_module! {
 			fund.raised  = fund.raised.checked_add(&value).ok_or(Error::<T>::Overflow)?;
 			ensure!(fund.raised <= fund.cap, Error::<T>::CapExceeded);
 
+			let old_balance = Self::contribution_get(index, &who);
```
Comment (shawntabrizi) [inline runtime/common/src/crowdloan.rs:337]: Should just put the users nonce?

File: runtime/common/src/crowdloan.rs [hunk 2]
```diff
@@ -334,9 +334,11 @@ decl_module! {
...
+
 			if let Some(ref verifier) = fund.verifier {
 				let signature = signature.ok_or(Error::<T>::InvalidSignature)?;
-				let payload = (index, value);
+				let payload = (index, &who, old_balance, value);
```
Comment (shawntabrizi) [inline runtime/common/src/crowdloan.rs:341]: Should just put the users nonce?

Comment (xlc) [reply runtime/common/src/crowdloan.rs:341]: there is no nonce for multisig / proxy

**Review by kianenigma — APPROVED** [review]

**Comment by shawntabrizi** [comment]
@xlc I guess this got closed because I merged my branch onto master. Feel free to reopen this against master and we can merge it pending you writing tests.

**Pull request closed without merging**